File: ios/LaneShadow/Views/Templates/RouteResultsScreen.swift
import Combine
import Foundation
import LaneShadowTheme
import SwiftUI

@MainActor
public protocol RouteResultsScreenInspectionSeam: AnyObject {
    var notice: PassthroughSubject<UInt, Never> { get }
    func visit(_ view: RouteResultsScreen, _ line: UInt)
}

/// RouteResultsScreen — displays three alternative route polylines with Navigator message and refine chat.
///
/// Composes `LSMapLayer` with three polylines, `LSNavigatorMessage` with three route attachment cards,
/// and `LSChatInput` for refinement. Data can come from `RouteResultsMockProvider` or a live state.
public struct RouteResultsScreen: View {
    @Environment(\.theme) private var theme

    private let state: RouteResultsScreenState
    private let onPin: @Sendable () -> Void
    private let onDismiss: @Sendable () -> Void
    private let onRouteCardTap: @Sendable (String) -> Void
    private let inspection: (any RouteResultsScreenInspectionSeam)?

    @State private var chatInputValue: String = ""
    @State private var drawProgress: [String: Double] = [:]
    @State private var isCalloutVisible: Bool = true

    public init(
        provider: RouteResultsMockProvider.Type = RouteResultsMockProvider.self,
        variant: String = "default",
        onPin: @escaping @Sendable () -> Void = {},
        onDismiss: @escaping @Sendable () -> Void = {},
        onRouteCardTap: @escaping @Sendable (String) -> Void = { _ in },
        inspection: (any RouteResultsScreenInspectionSeam)? = nil
    ) {
        self.init(
            state: provider.value(variant: variant),
            onPin: onPin,
            onDismiss: onDismiss,
            onRouteCardTap: onRouteCardTap,
            inspection: inspection
        )
    }

    init(
        state: RouteResultsScreenState,
        onPin: @escaping @Sendable () -> Void = {},
        onDismiss: @escaping @Sendable () -> Void = {},
        onRouteCardTap: @escaping @Sendable (String) -> Void = { _ in },
        inspection: (any RouteResultsScreenInspectionSeam)? = nil
    ) {
        self.state = state
        self.onPin = onPin
        self.onDismiss = onDismiss
        self.onRouteCardTap = onRouteCardTap
        self.inspection = inspection
    }

    public var body: some View {
        let content = baseBody

        if let inspection {
            content.onReceive(inspection.notice) { inspection.visit(self, $0) }
        } else {
            content
        }
    }

    private var baseBody: some View {
        ZStack(alignment: .topLeading) {
            navigatorMessageContainer
                .zIndex(2)

            LSMapLayer(
                map: {
                    mapView
                },
                bottomOverlays: [
                    GlassOverlaySlot(
                        id: "chatinput",
                        content: { chatInputView }
                    ),
                ],
                topBar: {
                    LSTopBar(
                        trailing: .none,
                        onMenuTap: {},
                        onNewTap: {}
                    )
                }
            )
            .accessibilityIdentifier("route-resultsscreen")
        }
        .onAppear {
            startRouteDrawAnimation()
        }
    }

    // MARK: - Map

    private var mapView: some View {
        LSMap(
            mode: .interactive,
            camera: CameraPosition(
                center: LatLng(lat: 37.7749, lon: -122.4194),
                zoom: 12
            ),
            cameraFit: .polylines(padding: .spacing4),
            polylines: routePolylines,
            annotations: routeAnnotations
        )
        .accessibilityIdentifier("maplayer.map")
    }

    private var routePolylines: [PolylineData] {
        state.routes.map { route in
            let progress = drawProgress[route.id] ?? 0.0
            let coordinates = decodePolyline(route.polyline)

            // Apply animation progress by trimming coordinates
            let animatedCoordinates: [LatLng]
            if progress < 1.0 {
                let count = max(1, Int(Double(coordinates.count) * progress))
                animatedCoordinates = Array(coordinates.prefix(count))
            } else {
                animatedCoordinates = coordinates
            }

            return PolylineData(
                coordinates: animatedCoordinates,
                variant: routeVariant(from: route.variant),
                strokeWidth: route.id == effectiveSelectedRouteId ? .lg : .md
            )
        }
    }

    private var routeAnnotations: [Annotation] {
        state.routes.compactMap { route in
            // Only show start/end markers for the selected route
            if route.id == effectiveSelectedRouteId {
                let coords = decodePolyline(route.polyline)
                if let first = coords.first, let last = coords.last {
                    return [
                        Annotation(kind: .start, coordinate: first, label: nil),
                        Annotation(kind: .end, coordinate: last, label: nil),
                    ]
                }
            }
            return nil
        }.flatMap { $0 }
    }

    private var effectiveSelectedRouteId: String? {
        state.selectedRouteId ?? state.routes.first?.id
    }

    private func routeVariant(from variant: String?) -> RouteVariant {
        guard let variant else { return .alt2 }
        switch variant {
        case "best":
            return .best
        case "alt1":
            return .alt1
        case "alt2":
            return .alt2
        default:
            return .alt2
        }
    }

    private func decodePolyline(_ encoded: String) -> [LatLng] {
        let decoded = decodeEncodedPolyline(encoded)
        return decoded.isEmpty ? fallbackCoordinates() : decoded
    }

    private func decodeEncodedPolyline(_ encoded: String, precision: Double = 5) -> [LatLng] {
        guard !encoded.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty else {
            return []
        }

        let factor = pow(10.0, precision)
        let bytes = Array(encoded.utf8)
        var index = 0
        var latitude = 0
        var longitude = 0
        var coordinates: [LatLng] = []

        func decodeValue() -> Int? {
            var result = 0
            var shift = 0

            while index < bytes.count {
                let byte = Int(bytes[index]) - 63
                index += 1
                result |= (byte & 0x1F) << shift
                shift += 5

                if byte < 0x20 {
                    return (result & 1) != 0 ? ~(result >> 1) : (result >> 1)
                }
            }

            return nil
        }

        while index < bytes.count {
            guard let deltaLat = decodeValue(),
                  let deltaLon = decodeValue()
            else {
                return []
            }

            latitude += deltaLat
            longitude += deltaLon

            coordinates.append(
                LatLng(
                    lat: Double(latitude) / factor,
                    lon: Double(longitude) / factor
                )
            )
        }

        return coordinates.count >= 2 ? coordinates : []
    }

    private func fallbackCoordinates() -> [LatLng] {
        [
            LatLng(lat: 37.7749, lon: -122.4194),
            LatLng(lat: 37.7849, lon: -122.4094),
            LatLng(lat: 37.7949, lon: -122.3994),
        ]
    }

    private func startRouteDrawAnimation() {
        // Reset draw progress
        drawProgress = Dictionary(uniqueKeysWithValues: state.routes.map { ($0.id, 0.0) })

        // Animate each route with stagger using theme motion tokens
        let staggerMs: Double = 120 // 120ms stagger between routes
        let durationMs = Double(theme.motion.duration["slower"] ?? 600)
        let easing = theme.motion.easing["standard"] ?? [0.4, 0, 0.2, 1]

        for (index, route) in state.routes.enumerated() {
            let delayMs = Double(index) * staggerMs

            DispatchQueue.main.asyncAfter(deadline: .now() + .milliseconds(Int(delayMs))) {
                withAnimation(
                    Animation.timingCurve(
                        easing[0],
                        easing[1],
                        easing[2],
                        easing[3],
                        duration: durationMs / 1000.0
                    )
                ) {
                    drawProgress[route.id] = 1.0
                }
            }
        }
    }

    // MARK: - Navigator Message Overlay

    private var navigatorMessageContainer: some View {
        VStack(alignment: .leading, spacing: 0) {
            navigatorMessageContent
        }
        .padding(.horizontal, theme.space.md)
        .padding(.top)
        .frame(maxWidth: .infinity, alignment: .topLeading)
        .accessibilityIdentifier("maplayer.topOverlay.navigator-message")
    }

    @ViewBuilder
    private var navigatorMessageContent: some View {
        if state.routes.isEmpty {
            emptyStateOverlay
        } else if isCalloutVisible {
            LSNavigatorMessage(
                body: state.message.body,
                attachments: routeAttachments,
                pinned: state.message.pinned,
                onPin: onPin,
                onDismiss: {
                    isCalloutVisible = false
                    onDismiss()
                },
                onRouteCardTap: { routeId in
                    onRouteCardTap(routeId)
                },
                selectedRouteId: effectiveSelectedRouteId
            )
        } else {
            recallChip
        }
    }

    private var routeAttachments: [LSRouteAttachment] {
        guard let attachments = state.message.attachments else { return [] }

        return attachments.compactMap { attachment in
            // Find the corresponding route
            guard let route = state.routes.first(where: { $0.id == attachment.routeId }) else {
                return nil
            }

            return LSRouteAttachment(
                id: attachment.routeId,
                label: route.name,
                description: route.via,
                distance: formatDistance(route.distance),
                duration: formatDuration(route.estimatedTime),
                scenicScore: Double(attachment.scenic),
                weatherBadge: LSWeatherBadgeInfo(
                    type: weatherBadgeType(from: attachment.weather.condition),
                    text: attachment.weather.label
                ),
                isBest: attachment.isBest
            )
        }
    }

    private var emptyStateOverlay: some View {
        LSEmptyState(
            title: "No routes available",
            body: "Try adjusting your start or end points."
        )
        .frame(maxWidth: .infinity)
    }

    private var recallChip: some View {
        Button(action: {
            isCalloutVisible = true
        }) {
            LSPill(size: .md) {
                LSText("Recall", variant: .label.md, color: .secondary)
            }
            .background(
                RoundedRectangle(cornerRadius: theme.radius.full, style: .continuous)
                    .fill(LaneShadowTheme.color.surface.card)
            )
            .overlay(
                RoundedRectangle(cornerRadius: theme.radius.full, style: .continuous)
                    .stroke(LaneShadowTheme.color.signal.default, lineWidth: theme.borderWidth.hairline)
            )
        }
        .buttonStyle(.plain)
        .accessibilityLabel("Recall route callout")
        .accessibilityIdentifier("routeresultsscreen-recall")
    }

    private func formatDistance(_ meters: Int) -> String {
        let miles = Double(meters) / 1609.34
        return String(format: "%.0f mi", miles)
    }

    private func formatDuration(_ seconds: Int) -> String {
        let hours = seconds / 3600
        let minutes = (seconds % 3600) / 60
        if hours > 0 {
            return String(format: "%dh %dm", hours, minutes)
        } else {
            return String(format: "%dm", minutes)
        }
    }

    private func weatherBadgeType(from condition: String) -> LSWeatherBadgeType {
        switch condition.lowercased() {
        case "clear":
            .clear
        case "rain":
            .rain
        case "wind":
            .wind
        case "storm":
            .cloudy
        default:
            .clear
        }
    }

    // MARK: - Chat Input

    private var chatInputView: some View {
        LSChatInput(
            value: $chatInputValue,
            placeholder: "Refine — 'make it shorter' / 'avoid Hwy 1'",
            onSend: { _ in },
            onCollapse: {},
            onFilter: {}
        )
        .padding(.horizontal, theme.space.md)
        .accessibilityIdentifier("route-resultsscreen-chatinput")
    }
}

// MARK: - Preview

#Preview {
    RouteResultsScreen()
}

File: android/app/src/main/java/com/laneshadow/ui/templates/RouteResultsScreen.kt
package com.laneshadow.ui.templates

import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.runtime.Composable
import androidx.compose.ui.Modifier
import androidx.compose.ui.platform.testTag
import com.laneshadow.sandbox.mockproviders.RouteResultsScreenState
import com.laneshadow.theme.LocalLaneShadowTheme
import com.laneshadow.ui.atoms.CameraPosition
import com.laneshadow.ui.atoms.CameraFit
import com.laneshadow.ui.atoms.LatLng
import com.laneshadow.ui.atoms.LSMap
import com.laneshadow.ui.atoms.MapMode
import com.laneshadow.ui.atoms.PolylineData
import com.laneshadow.ui.atoms.RouteVariant
import com.laneshadow.ui.atoms.SpacingToken
import com.laneshadow.ui.molecules.LSChatInput
import com.laneshadow.ui.molecules.LSRouteAttachmentCard
import com.laneshadow.ui.molecules.RouteAttachment
import com.laneshadow.ui.molecules.RouteAttachmentWeather
import com.laneshadow.ui.molecules.WeatherCondition
import com.laneshadow.ui.organisms.GlassOverlaySlot
import com.laneshadow.ui.organisms.LSMapLayer
import com.laneshadow.ui.organisms.LSNavigatorMessage
import com.laneshadow.ui.organisms.LSTopBar

/**
 * RouteResultsScreen template — three polylines + NavigatorMessage + refine chat.
 *
 * Renders the RouteResults screen with:
 * - Three concurrent polylines (best, alt1, alt2) with route-variant colors
 * - Pinned NavigatorMessage with three route attachment cards
 * - Camera auto-framing to union bounds with spacing.4 padding
 * - Refine chat input with placeholder
 *
 * Driven entirely by mock data from RouteResultsMockProvider — no live data fetching.
 *
 * @param state Screen state from RouteResultsMockProvider
 * @param onMenuTap Callback when hamburger menu is tapped
 * @param onRouteCardTap Callback when a route card is tapped
 * @param onPinTap Callback when pin icon is tapped
 * @param onDismissTap Callback when close icon is tapped
 * @param onRefineChange Callback when chat input value changes
 * @param onRefineSend Callback when send button is tapped
 * @param onCollapseTap Callback when collapse button is tapped
 * @param onFilterTap Callback when filter button is tapped
 * @param modifier Modifier for the root composable
 */
@Composable
fun RouteResultsScreen(
    state: RouteResultsScreenState,
    onMenuTap: () -> Unit,
    onRouteCardTap: (String) -> Unit,
    onPinTap: () -> Unit,
    onDismissTap: () -> Unit,
    onRefineChange: (String) -> Unit,
    onRefineSend: (String) -> Unit,
    onCollapseTap: () -> Unit,
    onFilterTap: () -> Unit,
    modifier: Modifier = Modifier,
) {
    val theme = LocalLaneShadowTheme.current

    // Convert mock routes to PolylineData list
    val polylines = state.routes.mapIndexed { index, route ->
        val variant = when (index) {
            0 -> RouteVariant.Best
            1 -> RouteVariant.Alt1
            else -> RouteVariant.Alt2
        }
        PolylineData(
            coordinates = emptyList(), // Mock data doesn't have real coordinates
            variant = variant,
        )
    }

    // Convert mock attachments to UI RouteAttachment list
    val attachments = state.message.attachments?.mapIndexed { index, mockAttachment ->
        val route = state.routes.find { it.id == mockAttachment.routeId }
        com.laneshadow.ui.molecules.RouteAttachment(
            id = mockAttachment.routeId,
            title = route?.name ?: "Route",
            via = route?.via ?: "",
            distance = formatDistance(route?.distance ?: 0),
            duration = formatDuration(route?.estimatedTime ?: 0),
            scenicScore = mockAttachment.scenic ?: 0,
            variant = when (mockAttachment.variant) {
                "best" -> RouteVariant.Best
                "alt1" -> RouteVariant.Alt1
                "alt2" -> RouteVariant.Alt2
                else -> RouteVariant.Best
            },
            weatherBadge = mockAttachment.weather?.let {
                RouteAttachmentWeather(
                    condition = when (it.condition) {
                        "clear" -> WeatherCondition.Clear
                        "rain" -> WeatherCondition.Rain
                        "wind" -> WeatherCondition.Wind
                        else -> WeatherCondition.Clear
                    },
                    label = it.label,
                )
            },
            isBest = mockAttachment.isBest == true,
        )
    } ?: emptyList()

    LSMapLayer(
        map = {
            LSMap(
                mode = MapMode.Preview,
                camera = CameraPosition(
                    center = LatLng(37.8104, -122.4752),
                    zoom = 11.0,
                ),
                cameraFit = CameraFit.Polylines(padding = SpacingToken.Spacing4),
                polylines = polylines,
            )
        },
        topOverlays = listOf(
            GlassOverlaySlot(
                id = "navigator-message",
                content = {
                    LSNavigatorMessage(
                        body = state.message.body,
                        attachments = attachments,
                        pinned = state.message.pinned == true,
                        onPin = onPinTap,
                        onDismiss = onDismissTap,
                        modifier = Modifier.testTag("route-results-navigator-message"),
                    )
                }
            )
        ),
        bottomOverlays = listOf(
            GlassOverlaySlot(
                id = "chat-input",
                content = {
                    LSChatInput(
                        value = "",
                        onValueChange = onRefineChange,
                        placeholder = "Refine — 'make it shorter' / 'avoid Hwy 1'",
                        onSend = onRefineSend,
                        onCollapse = onCollapseTap,
                        onFilter = onFilterTap,
                        modifier = Modifier.testTag("route-results-chat-input"),
                    )
                }
            )
        ),
        topBar = {
            LSTopBar(
                onMenuTap = onMenuTap,
                modifier = Modifier.testTag("route-results-topbar"),
            )
        },
        modifier = modifier.fillMaxSize(),
    )
}

private fun formatDistance(meters: Int): String {
    val miles = (meters / 1609.34).toInt()
    return "${miles} mi"
}

private fun formatDuration(seconds: Int): String {
    val hours = seconds / 3600
    val minutes = (seconds % 3600) / 60
    return if (hours > 0) {
        "${hours}h ${minutes}m"
    } else {
        "${minutes}m"
    }
}
